src/sagemaker_containers/environment.py

```diff
@@ -250,134 +241,265 @@ def __iter__(self):
         items = {_property: getattr(self, _property) for _property in self.properties()}
         return iter(items)
 
-    def __init__(self,
-                 input_dir,  # type: str
-                 input_config_dir,  # type: str
-                 model_dir,  # type: str
-                 output_dir,  # type: str
-                 hyperparameters,  # type: dict
-                 resource_config,  # type: dict
-                 input_data_config,  # type: dict
-                 output_data_dir,  # type: str
-                 hosts,  # type: () -> list
-                 channel_input_dirs,  # type: dict
-                 current_host,  # type: str
-                 num_gpu,  # type: int
-                 num_cpu,  # type: int
-                 module_name,  # type: str
-                 module_dir,  # type: str
-                 enable_metrics,  # type: bool
-                 log_level  # type: int
-                 ):
+    def __init__(self, session=None):
+
+        """
+        Args:
+            session (boto3.Session): an optional boto session to use if required.
+        """
+        session = session or boto3.Session()
```

Comment:
Also, I'm not sure we should fall back to using a Session to determine region at all. Having this variable populated is part of the SageMaker Training contract, isn't it? And therefore local mode will recreate this as well. It may be better to just fail instead of defaulting like this.

It depends on the use cases we want to support for these Env classes is. Are you supposed to be able to run them outside of a proper SageMaker environment and have them work? What's the expected behavior there?

This may also point to the need to have an init method with no side effects, and a separate factory method (.create()) which does all of the reading from the environment. If users want to unit test their code, this would let them do so without patching thins.